File: src/construction_mobile/lib/features/absences/data/models/absence.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'absence.freezed.dart';
part 'absence.g.dart';

/// Mirrors the API's `AbsenceDto`.
@freezed
abstract class Absence with _$Absence {
  const factory Absence({
    required String id,
    required String employeeId,
    required String employeeName,
    required String type,
    required String status,

    /// `YYYY-MM-DD`.
    required String startDate,

    /// `YYYY-MM-DD`, inclusive.
    required String endDate,

    /// Calendar days covered, both ends included.
    required int dayCount,
    String? reason,
    String? requestedByName,
    String? reviewedByName,
    DateTime? reviewedAt,
    String? reviewNote,

    /// `YYYY-MM-DD`. Set together with [proposedEndDate] while a change to
    /// this (already-approved) absence awaits confirmation from the other
    /// side.
    String? proposedStartDate,

    /// `YYYY-MM-DD`, inclusive.
    String? proposedEndDate,
    String? proposedReason,
    String? proposedByName,

    /// True when the employee proposed the change (so management must
    /// confirm it); false when management proposed it (so the employee
    /// must).
    @Default(false) bool proposedByEmployee,
    DateTime? proposedAt,
    required DateTime createdAt,
  }) = _Absence;

  const Absence._();

  DateTime? get start => DateTime.tryParse(startDate);

  DateTime? get end => DateTime.tryParse(endDate);

  DateTime? get proposedStart =>
      proposedStartDate == null ? null : DateTime.tryParse(proposedStartDate!);

  DateTime? get proposedEnd =>
      proposedEndDate == null ? null : DateTime.tryParse(proposedEndDate!);

  /// Nobody has answered it yet.
  bool get isPending => status == 'Requested';

  bool get isApproved => status == 'Approved';

  /// A change is staged and waiting on the other side to confirm or decline it.
  bool get hasPendingEdit => proposedStartDate != null;

  /// Whether the employee may take this one back.
  ///
  /// Only an unanswered request. Granted leave has to be refused by a
  /// supervisor instead — withdrawing it would leave work planned around days
  /// somebody is still away. The API enforces this; the flag exists so the
  /// screen does not offer a button that can only fail.
  bool get canWithdraw => isPending;

  factory Absence.fromJson(Map<String, dynamic> json) =>
      _$AbsenceFromJson(json);
}

File: src/Construction.Domain/Entities/Absence.cs
using Construction.Domain.Common;
using Construction.Domain.Enums;

namespace Construction.Domain.Entities;

/// <summary>
/// A stretch of days an employee is not at work.
/// </summary>
/// <remarks>
/// The other half of a schedule. Knowing who is posted to a site is only
/// useful alongside knowing who is actually going to turn up: a board that
/// shows a person on site during their annual leave is worse than no board,
/// because somebody plans around it.
///
/// Only an <see cref="AbsenceStatus.Approved"/> absence makes someone
/// unavailable. A request that has not been answered is a question, not a
/// fact, and the schedule must not act on it.
/// </remarks>
public class Absence : BaseEntity, ISoftDeletable, IAuditable
{
    public Guid EmployeeId { get; set; }

    public Employee Employee { get; set; } = null!;

    public AbsenceType Type { get; set; } = AbsenceType.AnnualLeave;

    public AbsenceStatus Status { get; set; } = AbsenceStatus.Requested;

    public DateOnly StartDate { get; set; }

    /// <summary>Inclusive: a single day off has the same start and end.</summary>
    public DateOnly EndDate { get; set; }

    public string? Reason { get; set; }

    public Guid? RequestedByUserId { get; set; }

    public User? RequestedByUser { get; set; }

    public Guid? ReviewedByUserId { get; set; }

    public User? ReviewedByUser { get; set; }

    public DateTime? ReviewedAt { get; set; }

    /// <summary>Why it was refused. Cleared when it is granted.</summary>
    public string? ReviewNote { get; set; }

    public bool IsDeleted { get; set; }

    public DateTime? DeletedAt { get; set; }

    /// <summary>
    /// A change to an already-approved absence's dates/reason, waiting on the
    /// other side to confirm it — see <c>ProposeAbsenceEditCommand</c> and
    /// <c>ConfirmAbsenceEditCommand</c>. Null fields mean no edit is pending;
    /// nothing here touches <see cref="StartDate"/>/<see cref="EndDate"/>
    /// until it is confirmed, so the schedule and leave balance never see a
    /// change nobody has agreed to yet.
    /// </summary>
    public DateOnly? ProposedStartDate { get; set; }

    public DateOnly? ProposedEndDate { get; set; }

    public string? ProposedReason { get; set; }

    public Guid? ProposedByUserId { get; set; }

    public User? ProposedByUser { get; set; }

    /// <summary>
    /// True when the employee themselves proposed the change (so it is
    /// waiting on management to confirm); false when management proposed it
    /// (so it is waiting on the employee).
    /// </summary>
    public bool ProposedByEmployee { get; set; }

    public DateTime? ProposedAt { get; set; }

    public bool HasPendingEdit => ProposedStartDate is not null;

    /// <summary>
    /// Whole days off, both ends included.
    /// </summary>
    public int DayCount => EndDate.DayNumber - StartDate.DayNumber + 1;

    /// <summary>True when this absence keeps the employee away on the given day.</summary>
    public bool CoversDay(DateOnly day) =>
        Status == AbsenceStatus.Approved && StartDate <= day && EndDate >= day;
}
